<a href="https://colab.research.google.com/github/SamuelCO1/talleresGoogleColab/blob/main/Taller_24_Branch_and_Bound.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## «*Dices que cada nueva mañana nos trae mil rosas sí, pero, ¿dónde están los pétalos de la rosa de ayer?*»
### [Omar Khayyam](https://es.wikipedia.org/wiki/Omar_Jayam)

# Taller 24 Branch and Bound



# Branch and Bound

Ejemplo de la solución  paso a paso de un problema de programación entera utilizando la heurística Branch and Bound.

## Problema inicial: problema00

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [2]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones00 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob00 = cvx.Problem(obj, restricciones00)

# Obtener el valor óptimo
prob00.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob00.status)
print("Valor máximo:", "%.2f"% prob00.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.25
Valor óptimo de x: 3.75
Valor óptimo de y: 2.25


## subproblema01

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45
   
   **x <= 3**

   x   > = 0      

   y   >=  0

**x, y enteros**


In [3]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones01 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x <= 3,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob01 = cvx.Problem(obj, restricciones01)

# Obtener el valor óptimo
prob01.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob01.status)
print("Valor máximo:", "%.2f"% prob01.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 39.00
Valor óptimo de x: 3.00
Valor óptimo de y: 3.00


## subproblema02

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

   y   >=  0

**x, y enteros**


In [4]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones02 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob02 = cvx.Problem(obj, restricciones02)

# Obtener el valor óptimo
prob02.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob02.status)
print("Valor máximo:", "%.2f"% prob02.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 41.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.80


## subproblema03

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      

  **y >= 2**

**x, y enteros**


In [5]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones03 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
y >= 2

# Restricciones de no negatividad

]

# Configurar el problema
prob03 = cvx.Problem(obj, restricciones03)

# Obtener el valor óptimo
prob03.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob03.status)
print("Valor máximo:", "%.2f"% prob03.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: infeasible
Valor máximo: -inf


TypeError: must be real number, not NoneType

## subproblema04

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [6]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones04 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob04 = cvx.Problem(obj, restricciones04)

# Obtener el valor óptimo
prob04.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob04.status)
print("Valor máximo:", "%.2f"% prob04.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.56
Valor óptimo de x: 4.44
Valor óptimo de y: 1.00


## subproblema05

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 5**      
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [7]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones05 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 5,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob05 = cvx.Problem(obj, restricciones05)

# Obtener el valor óptimo
prob05.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob05.status)
print("Valor máximo:", "%.2f"% prob05.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00


## subproblema06

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   **x >= 4**      
   
   **x <= 4**
   
   **y <= 1**

   y   >=  0

**x, y enteros**


In [9]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear la variable a optimizar
x = cvx.Variable()
y = cvx.Variable()

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones06 = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones subproblemas
x >= 4,
x <=4,
y <= 1,

# Restricciones de no negatividad
y >= 0
]

# Configurar el problema
prob06 = cvx.Problem(obj, restricciones06)

# Obtener el valor óptimo
prob06.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob06.status)
print("Valor máximo:", "%.2f"% prob06.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value)

Estado de la solución: optimal
Valor máximo: 37.00
Valor óptimo de x: 4.00
Valor óptimo de y: 1.00


# Resumen del Branch and Bound
## Problema inicial: problema00
Restricciones: x>=0; y>=0     

Resultados: valor óptimo: 41.25; x.value = 3.75; y.value = 2.25

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0     

Resultados: valor óptimo: 39.00; **x.value = 3.00**; **y.value = 3.00**

## subproblema02
Restricciones: **x>=4**; y>=0     

Resultados: valor óptimo: 41.00; x.value = 4.00; y.value = 1.80

## subproblema03
Restricciones: **x>=4**; **y >=2**     

Resultados: valor óptimo: infeasible; x.value = ???; y.value = ???

## subproblema04
Restricciones: **x>=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 40.56; x.value = 4.44; y.value = 1.00

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: valor óptimo: 40.00; **x.value = 5.00; y.value = 0.00**

## subproblema06
Restricciones: **x>=4**; **x<=4**; y>=0; **y <=1**     

Resultados: valor óptimo: 37.00; **x.value = 4.0; y.value = 1.0**

# Análisis de los resultados del Branch and Bound

Los subproblemas 01, 05 y 06 tienen una solución óptima factible (soluciones enteras).

## subproblema01
Restricciones: x>=0; **x<=3**; y>=0

Resultados: **valor óptimo: 39.00; x.value = 3.00; y.value = 3.00**

## subproblema05
Restricciones: **x>=5**; y>=0; **y <=1**

Resultados: **valor óptimo: 40.00; x.value = 4.0; y.value = 1.0**

## subproblema06
Restricciones: x>=4; x<=4; y>=0; y <=1

Resultados: **valor óptimo: 37.00; x.value = 4.0; y.value = 1.0**

Se selecciona como solución optima la del subproblema 05, ya que satisface las restricciones de enteros y genera el mayor valor óptimo.


# Problema inicial: problema00 resuelto directamente por CVXPY

Max Z = 8\* x + 5\* y

   x   +   y    <=  6
   
   9 \* x       + 5\* y        < =  45

   x   > = 0      

   y   >=  0

**x, y enteros**


In [8]:
# Importar la librería que se va a utilizar
import cvxpy as cvx

# Crear las variables a optimizar y establecerlas como enteros
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

# Definir la función objetivo para maximizar el ingreso
obj = cvx.Maximize(8*x + 5*y)

# Establecer las restricciones
restricciones_int = [

# Primera restricción
x + y <= 6,

# Segunda restricción
9*x + 5*y <= 45,

# Restricciones de no negatividad
x >= 0,
y >= 0
]

# Configurar el problema
prob_int = cvx.Problem(obj, restricciones_int)

# Obtener el valor óptimo
prob_int.solve()

# Mostrar el estado de la solución y el valor óptimo de la función objetivo
print("Estado de la solución:", prob_int.status)
print("Valor máximo:", "%.2f"% prob_int.value)

# Mostrar los valores óptimos de las variables de decisión
print ("Valor óptimo de x:", "%.2f"% x.value)
print("Valor óptimo de y:", "%.2f"% y.value, "\n")

# Soluciones con ECOS_BB
print ("Soluciones con ECOS_BB")

# Resolver con ECOS_bb
prob_int.solve(solver=cvx.ECOS_BB, mi_max_iters = 1000,
               mi_abs_eps= 1e-6, mi_rel_eps = 1e-3, verbose = True)

# prob_int.solve(solver=cvx.ECOS_BB, mi_max_iters = 1000,
#               mi_abs_eps= 1e-6, mi_rel_eps = 1e-3)

# Mostrar el estado de la solución  con ECOS_BB
print("Estado de la solución con ECOS_BB:", prob_int.status)

print("Valor óptimo de la función objetivo con ECOS_BB:", prob_int.value)

# Mostrar los valores óptimos de las variables de decisión con ECOS_BB
print("Valores óptimos de las variables de decisión con ECOS_BB:",
      x.value, y.value, "\n")

Estado de la solución: optimal
Valor máximo: 40.00
Valor óptimo de x: 5.00
Valor óptimo de y: 0.00 

Soluciones con ECOS_BB
                                     CVXPY                                     
                                     v1.3.3                                    
(CVXPY) Mar 19 01:33:47 PM: Your problem has 2 variables, 4 constraints, and 0 parameters.
(CVXPY) Mar 19 01:33:47 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Mar 19 01:33:47 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Mar 19 01:33:47 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Mar 19 01:33:47 PM: Compiling pro

# Ejercicios

## 00.
Resuelva con el método **Branch and Bound**, utilizando CVXPY, el ejemplo 9.2-1 del libro de (TAHA, 2012) página 336.

***Agregue todas las casillas de texto y de código que requiera***.

Max Z = 5x + 4y
x + y <= 5
10x + 6y <= 45
x>= 0
y>= 0

In [18]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 0,
y >= 0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  23.75
Los valores optimo de las variables son:  3.75 1.25


Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x>= 0, x<=3, y>= 0

In [20]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 0,
x<=3,
y >= 0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  23.00
Los valores optimo de las variables son:  3.00 2.00


Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x>= 4,  y>= 0

In [19]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 4,
y >= 0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  23.33
Los valores optimo de las variables son:  4.00 0.83


Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x>= 4, y = 0

In [25]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 4,
y >= 0,
y <=0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  22.50
Los valores optimo de las variables son:  4.50 0.00


Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x>= 4, y >= 1

In [23]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 4,
y >= 1
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x<= 4, y = 0

In [27]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x <= 4,
y >= 0,
y <=0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  20.00
Los valores optimo de las variables son:  4.00 0.00


Max Z = 5x + 4y, x + y <= 5, 10x + 6y <= 45, x>= 5, y = 0

In [26]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 5,
y >= 0,
y <=0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob0.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

## 01.
Resuelva directamente, utilizando CVXPY, el ejemplo 9.2-1 del libro de (TAHA, 2012) página 336.

In [12]:
import cvxpy as cvx
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

obj0 = cvx.Maximize(5*x + 4*y)

restri = [
x + y <= 5,
10*x + 6*y <= 45,
x >= 0,
y >= 0
]
prob0 = cvx.Problem(obj0, restri)
prob0.solve()
print('El estado de la solucion es: ', prob0.status)
print('El valor de la funcion objetivo es: ', prob0.value)
print(f'Los valores optimo de las variables son: {x.value} {y.value}')

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  23.0
Los valores optimo de las variables son: 3.0 2.0


## 02.
Resuelva con el método **Branch and Bound**, utilizando CVXPY, **uno** de los cinco numerales del ejercicio 2 del conjunto de problemas 9.2a del libro de (TAHA, 2012) página 342.

***Agregue todas las casillas de texto y de código que requiera***.

Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, y>=0

In [30]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
y >= 0
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.47
Los valores optimo de las variables son:  3.71 2.35


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=3, y>=0

In [31]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=3,
y >= 0
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.00
Los valores optimo de las variables son:  3.00 2.67


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=3, y>=0, y<=2 **ESTE DIO**

In [33]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=3,
y >= 0,
y<=2
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  12.00
Los valores optimo de las variables son:  3.00 2.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=3, y>=3

In [34]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=3,
y >= 3
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  13.50
Los valores optimo de las variables son:  2.25 3.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=2, y>=3

In [38]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=2,
y >= 3
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  13.33
Los valores optimo de las variables son:  2.00 3.11


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=2, y<=3 **ESTE DIO**

In [40]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=2,
y <= 3
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  13.00
Los valores optimo de las variables son:  2.00 3.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=2, y>=4 **ESTE DIO**

In [39]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
x<=2,
y >= 4
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  12.00
Los valores optimo de las variables son:  0.00 4.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=3, y>=3 **ESTE NO DIO**

In [37]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 3,
y >= 3
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=4, y>=0

In [32]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 4,
y >= 0
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.43
Los valores optimo de las variables son:  4.00 2.14


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=4, y>=0, y<= 2

In [35]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 4,
y >= 0,
y<=2
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.40
Los valores optimo de las variables son:  4.20 2.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, x<=4, y>=0, y<= 2 **ESTE DIO**

In [41]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x>=0,
x <= 4,
y >= 0,
y<=2
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.00
Los valores optimo de las variables son:  4.00 2.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=5, y>=0, y<= 2

In [42]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 5,
y >= 0,
y<=2
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.29
Los valores optimo de las variables son:  5.00 1.43


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=5, y>=0, y<= 1

In [43]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 5,
y >= 0,
y<=1
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.20
Los valores optimo de las variables son:  5.60 1.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, X>=0, x<=5, y>=0, y<= 1 **ESTE DIO**

In [46]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x>=0,
x <= 5,
y >= 0,
y<=1
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  13.00
Los valores optimo de las variables son:  5.00 1.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=6, y>=0, y<= 1

In [47]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 6,
y >= 0,
y<=1
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.14
Los valores optimo de las variables son:  6.00 0.71


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=6, y>=0, y<= 0 **ESTE DIO**

In [48]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 6,
y >= 0,
y<=0
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.00
Los valores optimo de las variables son:  7.00 0.00


Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=6, y>=1 **ESTE DIO**

In [49]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 6,
y >= 1
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=5, y>=2 **ESTE NO DIO**

In [44]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 5,
y >= 2
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=4, y>=3 **ESTE NO DIO**

In [36]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable()
y = cvx.Variable()

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 4,
y >= 3
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  infeasible
El valor de la funcion objetivo es:  -inf


TypeError: must be real number, not NoneType

## 03.
Resuelva directamente, utilizando CVXPY, **uno** de los cinco numerales del ejercicio 2 del conjunto de problemas 9.2a del libro de (TAHA, 2012) página 342.

Sugerencia: utilice el mismo numeral utilizado en el ejercicio 03.

Max Z = 2x + 3y, 5x + 7y <= 35, 4x + 9y <=36, x>=0, y>=0

In [50]:
# Resuelva el ejercicio a partir de aquí.
import cvxpy as cvx
x = cvx.Variable(integer=True)
y = cvx.Variable(integer=True)

obj1 = cvx.Maximize(2*x + 3*y)

restri1 = [
5*x + 7*y <= 35,
4*x + 9*y <=36,
x >= 0,
y >= 0
]
prob1 = cvx.Problem(obj1, restri1)
prob1.solve()
print('El estado de la solucion es: ', prob1.status)
print('El valor de la funcion objetivo es: ', '%.2f'% prob1.value)
print('Los valores optimo de las variables son: ', '%.2f'% x.value, '%.2f'% y.value)

El estado de la solucion es:  optimal
El valor de la funcion objetivo es:  14.00
Los valores optimo de las variables son:  4.00 2.00
